File: HekimBuradaUI/src/lib/presenceHub.ts
import * as signalR from "@microsoft/signalr";
import { MESSAGING_URL } from "./api";
import { auth } from "./auth";

export interface LiveNotification {
  title: string;
  body: string;
  linkPath: string;
}

/** Navbar'ın kurduğu tek presence bağlantısından, o an açık olan HERHANGİ bir sayfaya "veri bayatladı,
 * kendini yenile" sinyali — sayfa bileşenleri bunu dinleyip kendi load()/loadAll() fonksiyonlarını
 * tekrar çağırır (bkz. proje kararı: aynı sayfadayken F5 gerekmesin). */
export const LIVE_NOTIFICATION_EVENT = "hekimburada:notification";

/**
 * Uygulama genelinde (her sayfada, sadece sohbet ekranında değil) bağlı kalınan presence bağlantısı —
 * kullanıcının "çevrimiçi" sayılmasını sağlar (bkz. Messaging/Hubs/PresenceHub.cs) ve teklif/mesaj
 * gibi olaylarda anlık push bildirimlerini alır. Token yoksa/çıkış yapılırsa bağlanmaz.
 */
export function connectPresence(onNotification: (notification: LiveNotification) => void) {
  const token = auth.getToken();
  if (!token) {
    return () => {};
  }

  const connection = new signalR.HubConnectionBuilder()
    .withUrl(`${MESSAGING_URL}/hubs/presence?access_token=${token}`)
    .withAutomaticReconnect()
    .build();

  connection.on("notificationReceived", (notification: LiveNotification) => {
    onNotification(notification);
    window.dispatchEvent(new CustomEvent<LiveNotification>(LIVE_NOTIFICATION_EVENT, { detail: notification }));
  });
  connection.start().catch(() => {});

  return () => {
    connection.stop();
  };
}

File: services/HekimBurada.Messaging/Hubs/PresenceHub.cs
using System.Security.Claims;
using Messaging.Presence;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.SignalR;

namespace Messaging.Hubs;

/// <summary>
/// Uygulama genelinde (sadece sohbet ekranında değil, her sayfada) bağlı kalınan hub — kullanıcının
/// "çevrimiçi" sayılmasını sağlar ve PresenceGrpcService'in anlık bildirim push'ları için hedef
/// olur (bkz. proje kararı: teklif/mesaj bildirimlerinde SignalR mi e-posta mı gönderileceğine
/// bu bağlantının varlığına göre karar verilir). CodeGen dışı, elle eklendi.
/// </summary>
[Authorize]
public sealed class PresenceHub : Hub
{
    private readonly IPresenceTracker _tracker;

    public PresenceHub(IPresenceTracker tracker)
    {
        _tracker = tracker;
    }

    public override async Task OnConnectedAsync()
    {
        var userId = GetUserId();
        if (userId is not null)
        {
            await _tracker.ConnectAsync(userId.Value);
        }

        await base.OnConnectedAsync();
    }

    public override async Task OnDisconnectedAsync(Exception? exception)
    {
        var userId = GetUserId();
        if (userId is not null)
        {
            await _tracker.DisconnectAsync(userId.Value);
        }

        await base.OnDisconnectedAsync(exception);
    }

    private Guid? GetUserId()
    {
        var raw = Context.User?.FindFirst("sub")?.Value ?? Context.User?.FindFirst(ClaimTypes.NameIdentifier)?.Value;
        return Guid.TryParse(raw, out var id) ? id : null;
    }
}
